debian/gndaccess.1
.TH "GNDACCESS" "1" "" "Manual" ""
.SH NAME
.PP
gndaccess \- Access GND records via HTTP
.SH SYNOPSIS
.PP
The application is automatically started as service, listening on port
6699.
.IP
.nf
\f[C]
sudo\ service\ gndaccess\ {status|start|stop|restart}
\f[]
.fi
.SH DESCRIPTION
.PP
This applications provices a web service to access GND records via HTTP
in different formats.
The format must be provided with URL query parameter \f[C]format\f[].
.IP \[bu] 2
<http://localhost:6699/4021477-1?format=aref>
.SS Supported formats
.IP \[bu] 2
\f[C]aref\f[]: aREF format (RDF in JSON for easy access, similar to
JSON\-LD)
.IP \[bu] 2
\f[C]nt\f[]: NTriples
.IP \[bu] 2
\f[C]rdfxml\f[]: RDF/XML
.IP \[bu] 2
\f[C]marcxml\f[]
.SS CORS and JSONP
.PP
Cross\-Origin Resource Sharing (CORS) is enabled by default.
In addition all JSON\-based formats can can be returned in JSONP with
parameter \f[C]callback\f[]:
.IP \[bu] 2
<http://localhost:6699/4021477-1?format=aref&callback=abc>
.SS HTML client
.PP
The application includes an HTML client, consisting of the following
files:
.IP \[bu] 2
\f[C]index.html\f[]
.IP \[bu] 2
\f[C]gndaccess.css\f[]
.IP \[bu] 2
\f[C]gndaccess.json\f[]
.IP \[bu] 2
\f[C]formats.json\f[] (generated dynamically)
.SH INSTALLATION
.PP
The application is packaged as Debian package.
Releases can be found at
.IP \[bu] 2
<https://github.com/gbv/gndaccess/releases>
.PP
No binaries are included, so the package should work on all
architectures.
It is tested with Ubuntu 12.04 LTS and Ubuntu 14.04 LTS.
.PP
Files are installed at the following locations:
.IP \[bu] 2
\f[C]/srv/gndaccess/\f[] \- application
.IP \[bu] 2
\f[C]/var/log/gndaccess/\f[] \- log files \f[C]access.log\f[] and
\f[C]error.log\f[]
.IP \[bu] 2
\f[C]/etc/logrotate.d/gndaccess\f[] \- logrotate configuration (compress
daily, keep 14 days)
.IP \[bu] 2
\f[C]/etc/default/gndaccess\f[] \- server configuration
.IP \[bu] 2
\f[C]/etc/gndaccess/\f[] \- application configuration
.SH CONFIGURATION
.PP
See \f[C]/etc/default/gndaccess\f[] for basic configuration.
Settings are not modified by updates.
Only simple key value\-pairs are allowed with the following keys:
.IP \[bu] 2
\f[C]PORT\f[] \- port number (required, 6699 by default)
.IP \[bu] 2
\f[C]WORKERS\f[] \- number of parallel connections (required, 5 by
default).
If put behind a HTTP proxy, this number is not affected by slow cient
connections but only by the time of processing each request.
.IP \[bu] 2
\f[C]PROXY\f[] \- a space\-or\-comma\-separated list of trusted IPs or
IP\-ranges (e.g.
\f[C]192.168.0.0/16\f[]) to take from the \f[C]X\-Forwarded\-For\f[]
header.
The special value \f[C]*\f[] can be used to trust all IPs.
.PP
Restart is needed after changes.
.PP
The HTML client can be customized by putting static files into directory
\f[C]/etc/gndaccess/htdocs\f[] to override files in
\f[C]/srv/gndaccess/htdocs\f[].
.SH SEE ALSO
.PP
Changelog is located in file
\f[C]debian/changelog\f[] (debian/changelog) in the source code
repository.
.PP
Source code and issue tracker at <https://github.com/gbv/gndaccess>.
See file \f[C]CONTRIBUTING.md\f[] source code organization.
.PP
See <http://www.dnb.de/entityfacts> for a similar service.


CHANGES.md
# 0.0.5

* include MARCXML format


lib/GBV/App/GNDAccess.pm
package GBV::App::GNDAccess;
use v5.14.1;

our $VERSION="0.0.5";
our $NAME="gndaccess";

use RDF::aREF;
use RDF::Trine;
use RDF::Trine::Serializer;
use JSON;
use Unicode::Normalize;
use Plack::Builder;
use HTTP::Tiny;
use Plack::Request;
use parent 'Plack::Component';

sub formats {
    return {
        aref => { 
            name => 'aREF', 
            type => 'application/json', 
            rdf  => 'aref',
        },
        html => { 
            name => 'HTML', 
            type => 'text/html' 
        },
        nt => { 
            name => 'NTriples', 
            type => 'application/n-triples',
            rdf  => 'NTriples',
        },
        rdfxml => { 
            name => 'RDF/XML', 
            type => 'application/rdf+xml',
            rdf  => 'RDFXML',
        },
        marcxml => {
            name => 'MARCXML',
            type => 'application/marcxml+xml',
            marc => 'marcxml'
        }
    }
}

sub config {
    my ($self, $file) = @_;
    return unless -f $file;

    local(@ARGV) = $file; 
    while(<>) {
        next if $_ !~ /^\s*([A-Z][_A-Z0-9]*)\s*=\s*([^#\n]*)/;
        $self->{$1} = $2;
        if ($1 eq 'PROXY' and $2 !~ /^[*]\s*$/) {
            $self->{TRUST} = [ split /\s*,\s*|\s+/, $self->{PROXY} ];
        }
    }
}    

sub prepare_app {
    my ($self) = @_;
    return if $self->{app};

    # load config file
    $self->config( grep { -f $_ } "debian/$NAME.default", "/etc/default/$NAME" );

    # build middleware stack
    $self->{app} = builder {
        enable_if { $self->{PROXY} } 'XForwardedFor', trust => $self->{TRUST};
        enable 'CrossOrigin', origins => '*';
        enable 'Rewrite', rules => sub {
            s{^/$}{/index.html}; return
        };
        enable 'Static', path => qr{\.(html|js|ico|css|png)},
            pass_through => 1, root => "/etc/$NAME/htdocs";
        enable 'Static', path => qr{\.(html|js|ico|css|png)},
            pass_through => 1, root => './htdocs';
        enable 'ContentLength';
        enable 'JSONP';
        enable 'Negotiate', 
            parameter => 'format',
            formats => $self->formats;
        builder {
            mount '/formats.json' => sub {
                $self->json(200,$self->formats);
            };
            mount '/' => sub {
                $self->main(@_) 
            };
        }
    }; 
}

sub call {
    my ($self, $env) = @_;
    $self->{app}->($env);
}

sub json {
    my ($self, $code, $data) = @_;
    my $json = JSON->new->utf8->pretty->encode($data);
    return [$code, ['Content-Type' => 'application/json; encoding=UTF-8'], [$json]];
}

sub main { 
    my ($self, $env) = @_;
    my $req = Plack::Request->new($env);

    my $path = substr($req->path,1) || ''; 
    if ( $path !~ qr{^(http://d-nb\.info/gnd/)?([0-9X-]+)$}) {
        return [404, ['Content-Type' => 'text/plain'], ['invalid GND identifier'. $path]];
    }
    my ($uri,$id) = ("http://d-nb.info/gnd/$2",$2);

    my $format = $env->{'negotiate.format'} || 'html';

    unless ($format = $self->formats->{$format}) {
        return [400, ['Content-Type' => 'text/plain'], ['format not supported']];
    }

    # RDF-based formats
    if ($format->{rdf}) {
        my $model = eval { get_rdf_NFKC($uri) };
        if ($@ || !$model || !$model->size) {
            return $self->json(404 => { error => "GND not found via $uri" });
        }

        if ($format->{rdf} == 'aref') {
            my $aref = encode_aref $model;
            return $self->json( 200, $aref );
        } elsif ($format->{rdf}) {
            use Encode qw(decode_utf8);
            my $rdf = RDF::Trine::Serializer->new($format->{rdf})
                    ->serialize_model_to_string($model);
            return [200,[], [decode_utf8($rdf)]];
        }

    # MARC-based formats
    } elsif ($format->{marc}) {
        my $url = "http://d-nb.info/gnd/$id/about/marcxml";
        my $res = HTTP::Tiny->new->get($url) || {};
        my $marcxml = $res->{content};
        unless ( $res->{success} and length $marcxml) {
            return $self->json(404 => { error => "GND not found via $url" });
        }
        $marcxml =~ s/^\s+//m;
        if ($marcxml !~ qr{^<\?xml}) {
            $marcxml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n$marcxml";
        }

        if ($format->{marc} eq 'marcxml') {
            return [200,['Content-Type' => 'application/marcxml+xml'],[$marcxml]];
        }
    }

    # html
    return [200, ['Content-Type' => 'text/plain'], ['GND gefunden (probier mal format=aref!)']];
}

sub get_rdf_NFKC {
    my ($uri) = @_;
    my $model = RDF::Trine::Model->new;
    $model->begin_bulk_ops;
    RDF::Trine::Parser->parse_url( $uri, sub {
        my $s = $_[0];
        if ($s->object->is_literal) {
            my $value = Unicode::Normalize::normalize('KC', $s->object->literal_value);
            $s->object->literal_value($value);
        }
        $model->add_statement($s); 
    } );
    $model->end_bulk_ops;
    return $model;
}

1;
